### PR TITLE
Mapped forcefield matrix + minor spelling changes 

### PR DESCRIPTION
**Pull request by olivier-goury**
To experiment a new handling of sparse matrices under mappings in our plugin, a few changes are needed in the core of SOFA.
Main changes:
- Add a new function in MechanicalObject (commit 6a6fd98).
- Refactoring of MatrixDerivId: warning holonomicC() is deprecated --> replaced by constraintMatrix (commit 966f9af).
- Change allocation in createMatrix from a fullMatrix to a sparseMatrix.
Linked to #26 

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
Hi Olivier,

thank for your PR. Let's see of it build [ci-build]. 
I let other person to handle this PR as we are in the same team :)

**Comment by damienmarchal** [comment]
Hi @olivier-goury , @ChristianDuriez 

Some tests on our CI are crashing with your PR, following the link 
https://www.sofa-framework.org/dash/?branch=pr/mapped_forcefield_matrix
You will find the crashing tests, those are the one that have a negative runtime, in the stackstrace you can see it is related to some Eigen spare matrix operation.

File: SofaKernel/framework/sofa/core/VecId.h [hunk 1]
```diff
@@ -150,8 +151,14 @@ class TStandardVec<V_MATDERIV, vaccess>
 public:
     typedef TVecId<V_MATDERIV, vaccess> MyVecId;
 
-    static MyVecId holonomicC()    { return MyVecId(1);}
-    static MyVecId nonHolonomicC() { return MyVecId(2);}
+    static MyVecId constraintMatrix()    { return MyVecId(1);} // jacobian matrix of constraints
```
Comment (matthieu-nesme) [inline SofaKernel/framework/sofa/core/VecId.h]: The matrix names are really unclear. Would it be possible to have some doc about it?
'constraintMatrix' seems to be 'constraintJacobian'? Does that mean that only the mstates in the same node as a Constraint would have such a matrix?
Whereas 'mappingMatrix' should be 'sparseJacobian', and every mstates should got one?

File: SofaKernel/framework/sofa/core/behavior/BaseMechanicalState.h [hunk 2]
```diff
@@ -226,6 +226,9 @@ class SOFA_CORE_API BaseMechanicalState : public virtual BaseState
     /// build the jacobian of the constraint in a baseMatrix
     virtual void getConstraintJacobian(const ExecParams* params, sofa::defaulttype::BaseMatrix* J,unsigned int & off) = 0;
 
+    /// fill the jacobian matrix (of the constraints) with identity blocks on the provided list of nodes(dofs)
```
Comment (matthieu-nesme) [inline SofaKernel/framework/sofa/core/behavior/BaseMechanicalState.h]: For the usage I have in mind, it would be more useful to start building the sparse Jacobians from ForceFields/Constraints. I am not sure of the interest of starting from a full Identity.
Starting from a selection matrix (identity with 0s) corresponding to the dofs where the ForceFields/Constraints are contributing allows to have the as parse as possible Jacobians, possibly null for branches where there is no ForceFields/Constraints.
Note building such a selection matrix is actually managed by `BaseForceField::updateForceMask`.

File: SofaKernel/modules/SofaBaseMechanics/UniformMass.inl [hunk 3]
```diff
@@ -75,13 +75,13 @@ using defaulttype::BaseMatrix;
 template <class DataTypes, class MassType>
```
Comment (matthieu-nesme) [inline SofaKernel/modules/SofaBaseMechanics/UniformMass.inl:75]: This file is not really related to the PR, is it?

File: SofaKernel/modules/SofaBaseLinearSolver/CompressedRowSparseMatrix.h [hunk 4]
```diff
@@ -174,6 +174,11 @@ public :
     {
     }
 
+    ~CompressedRowSparseMatrix()
```
Comment (matthieu-nesme) [inline SofaKernel/modules/SofaBaseLinearSolver/CompressedRowSparseMatrix.h:177]: This file is not really related to the PR, is it?

File: SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl [hunk 5]
```diff
@@ -2623,6 +2623,32 @@ void MechanicalObject<DataTypes>::getConstraintJacobian(const core::ExecParams*
 }
 
 template <class DataTypes>
+void MechanicalObject<DataTypes>::buildIdentityBlocksInJacobian(const sofa::helper::vector<unsigned int>& list_n, core::MatrixDerivId &mID)
```
Comment (matthieu-nesme) [inline SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl]: To me, it breaks the purpose of mstates to make them build matrices...
To me mstates should only store state vectors, but this idea seems definitively lost.

File: modules/SofaConstraint/LCPConstraintSolver.h [hunk 6]
```diff
@@ -168,6 +168,8 @@ class SOFA_CONSTRAINT_API LCPConstraintSolver : public ConstraintSolverImpl
     Data<double> showCellWidth;
     Data<defaulttype::Vector3> showTranslation;
     Data<defaulttype::Vector3> showLevelTranslation;
+    Data< sofa::helper::vector<double> > delta;
```
Comment (matthieu-nesme) [inline modules/SofaConstraint/LCPConstraintSolver.h]: are these new Data use somewhere?

File: modules/SofaMiscMapping/CMakeLists.txt [hunk 7]
```diff
@@ -71,7 +71,12 @@ set(SOURCE_FILES
 
 add_library(${PROJECT_NAME} SHARED ${HEADER_FILES} ${SOURCE_FILES})
 target_link_libraries(${PROJECT_NAME} PUBLIC SofaSimpleFem SofaRigid)
+
+if(USE_COMPRESSED_ROW_SPARSE)
```
Comment (matthieu-nesme) [inline modules/SofaMiscMapping/CMakeLists.txt]: OK to have a new compilation variable 'USE_COMPRESSED_ROW_SPARSE'.
Indeed it is problematic, that we want the same code, with the same API, described with different matrix structure, and I have no elegant proposition to handle that.
But where would be defined USE_COMPRESSED_ROW_SPARSE, you defined it manually in your cmakecache?
To be clearer, would it be possible to simply add 
`add_definitions("-DUSE_COMPRESSED_ROW_SPARSE")` rather than duplicating code.

File: modules/SofaMiscMapping/SubsetMultiMapping.inl [hunk 8]
```diff
@@ -28,6 +28,7 @@
 #include <iostream>
 #include <SofaBaseMechanics/IdentityMapping.h>
 
+#define USE_COMPRESSED_ROW_SPARSE 1
```
Comment (matthieu-nesme) [inline modules/SofaMiscMapping/SubsetMultiMapping.inl]: DEFINITIVELY NOT in the code!

**Review by matthieu-nesme — COMMENTED** [review]
The visitor building sparse Jacobians is missing, and I still do not really get the point of these partially sparse Jacobians.
But despite the matrix vectors renaming, I do not see anything breaking the API, so it should not create troubles.

File: modules/SofaMiscMapping/CMakeLists.txt (hunk 7 above)
Comment (ChristianDuriez) [reply modules/SofaMiscMapping/CMakeLists.txt]: My bad...  It was an intermediate code that was useful to get getJs but we no longer use this function
This will be removed from the pull request

File: modules/SofaMiscMapping/SubsetMultiMapping.inl (hunk 8 above)
Comment (ChristianDuriez) [reply modules/SofaMiscMapping/SubsetMultiMapping.inl]: Same comment

File: SofaKernel/modules/SofaBaseLinearSolver/CompressedRowSparseMatrix.h (hunk 4 above)
Comment (ChristianDuriez) [reply SofaKernel/modules/SofaBaseLinearSolver/CompressedRowSparseMatrix.h:177]: you are right... but It is maybe useful to have a destructor for CompressedRowSparseMatrix that does a clear() before destruction.

File: SofaKernel/framework/sofa/core/behavior/BaseMechanicalState.h (hunk 2 above)
Comment (ChristianDuriez) [reply SofaKernel/framework/sofa/core/behavior/BaseMechanicalState.h]: Here the Identity is sparse and not necessary full (as you have a list of blocks). 
It is implemented in the Mechanical State because these Jacobians are stored in the Mechanical State. But this function can be called, of course from ForceFields / Constraints

File: SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl (hunk 5 above)
Comment (ChristianDuriez) [reply SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl]: We can remove Jacobians from Mechanical States and store them in an other component. 
But this would be an other modification of SOFA core.

File: SofaKernel/framework/sofa/core/VecId.h (hunk 1 above)
Comment (ChristianDuriez) [reply SofaKernel/framework/sofa/core/VecId.h]: Yes we can rename constraintMatrix by constraintJacobian it is perhaps more clear.
In this case, we could name the second mappingJacobian (to be coherent).

File: modules/SofaConstraint/LCPConstraintSolver.h (hunk 6 above)
Comment (ChristianDuriez) [reply modules/SofaConstraint/LCPConstraintSolver.h]: Thank you for your comment... 
My bad, again something that should not be included in the Pull Request

**Comment by olivier-goury** [comment]
Changes were made according to the comments.

**Comment by olivier-goury** [comment]
Thank you for your comments

**Comment by damienmarchal** [comment]
Are the change ok ?
I rebuild [ci-build] the change against the current master.

**Comment by damienmarchal** [comment]
There is few new warning during the compilation process thats need to be fixed. 

These warnings are really minor details but I think it is important to do our best to decrease the general number of trivial warnings so that we can more efficiently detect the non-trivial one.

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by olivier-goury** [comment]
Thank you, warnings removed!

**Comment by olivier-goury** [comment]
We took care of the comments, would you like to merge this PR or do you see other issues?
Thanks

**Comment by matthieu-nesme** [comment]
Please, before merging, I'd like to get the interest of this PR?

Like already said, it does not bring anything but break the API, and polluate BaseMechanicalState with a off-purpose virtual function. So better motivate these modifications before merging.

**Comment by ChristianDuriez** [comment]
Thank you for your comment
The interest is to be able to map the forcefields through the mapping in a sparse manner for many solvers of SOFA.

I agree that this functionality is not directly available in the API and for now, would be only available in a private plug-in. But this is work in progress and we need more time to have a clean code available.

Why does it break the API ?

**Comment by olivier-goury** [comment]
We created a new tag SOFA_WITH_EXPERIMENTAL_FEATURES, set to OFF by default, which defines if the function buildIdentityBlocksInJacobian is used.
Hopefully, this PR can be merged now?

**Comment by guparan** [comment]
For some reason the latest commits have been ignored by the CI. 
Let's [ci-build] them.

**Comment by bcarrez** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
[ci-build]
